File: leetcode-js/016-3sum-closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    if (nums.length < 3) {
        return [];
    }

    nums.sort((a, b) => a - b);

    let minDiff = Infinity;

    for (let i = 0; i < nums.length; i += 1) {
        let left = i + 1;
        let right = nums.length - 1;

        while (left < right) {
            const diff = nums[i] + nums[left] + nums[right] - target;
            if (Math.abs(diff) < Math.abs(minDiff)) {
                minDiff = diff;
            }
            if (diff === 0) {
                return target;
            } else if (diff < 0) {
                left += 1;
            } else {
                right -= 1;
            }
        }
    }

    return target + minDiff;
};

File: leetcode-java/016-3sum-closest.java
public class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);

        int closestSum = nums[0] + nums[1] + nums[nums.length - 1];

        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            int left = i + 1;
            int right = nums.length - 1;

            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum < target) {
                    //move closer to target sum.
                    while (left < right && nums[left] == nums[left + 1]) {
                        left++;
                    }
                    left++;
                } else if (sum > target) {
                    while (left < right && nums[right] == nums[right - 1]) {
                        right--;
                    }
                    right--;
                } else {
                    return sum;
                }
                if (Math.abs(target - sum) < Math.abs(target - closestSum)) {
                    closestSum = sum;
                }
            }
        }

        return closestSum;
    }
}
